[stdlib] Add `repr` support for `List`, `Deque`, `Dict`, `LinkedList`, `Optional`, `Set`.

rd4com opened the pull request:
Hello,
this adds an overload to `repr` to support it. (redo of PR 4267, ping to @laszlokindrat )

It is a workaround that has to be replaced one day,
but seem like something widely used by users !

```mojo
 print(repr(List[UInt8](0, 1)))
 # [SIMD[DType.uint8, 1](0), SIMD[DType.uint8, 1](1)]
```
Also work with `List[DType]` for metaprogramming:
```mojo
def main():
    x = List(DType.int8, DType.int16)
    print(repr(x))
    print(DType.int8 in x)
```

rd4com committed:
[stdlib] Add `repr` support for `List[Movable&Copyable&Representable]`.

Hello,
this adds an overload to `repr` to support this.

It is a workaround that has to be replaced one day,
but seem like something widely used by users !

```mojo
 print(repr(List[UInt8](0, 1)))
 # [SIMD[DType.uint8, 1](0), SIMD[DType.uint8, 1](1)]
```

Signed-off-by: rd4com <144297616+rd4com@users.noreply.github.com>

rd4com committed:
Apply mojo format

JoeLoser reviewed (approved):
LGTM, thanks!  Should we also add future convenient wrapper `repr` function support for `Dict`, `Set`, and friends until we have better conditional conformances?  This may future-proof some "why does this work on `List` but not <xyz> common core collection" questions.

rd4com committed:
[stdlib] Add `repr` support for `List`, `Deque`, `Dict`, `LinkedList`, `Optional`, `Set`.

Hello,
this adds an overload to repr to support it. (redo of PR 4267, ping to @laszlokindrat )

It is a workaround that has to be replaced one day,
but seem like something widely used by users !

 print(repr(List[UInt8](0, 1)))
 # [SIMD[DType.uint8, 1](0), SIMD[DType.uint8, 1](1)]

Also work with List[DType] for metaprogramming:

def main():
    x = List(DType.int8, DType.int16)
    print(repr(x))
    print(DType.int8 in x)

rd4com committed:
Add entry to changelog.

laszlokindrat commented on mojo/stdlib/test/collections/test_optional.mojo:
```diff
@@ -186,6 +186,13 @@ def test_optional_unwrap():
 
+def test_optional_repr_wrap():
+    var o = Optional(10)
+    assert_equal(repr(o), "Optional(10)")
+    o = None
+    assert_equal(repr(o), "Optional(None)")
```
Unrelated to this PR, but this is a bit confusing. `Optional(10)` should probably just print `10`, since that's the actual value (and similarly, the second test should yield the same as `repr(None)`. The current implementation is kinda weird, since it prints `Optional`, but not the underlying type.

rd4com replied:
Yep, `List` for example do give the type :+1:

laszlokindrat replied:
If you want, seems like fixing this could be a good followup :)

rd4com replied:
Sure, let's go about fixing this (followup)! 
Here is an overview of what types do on `repr`:

Collections type do not always print their own type, some hints at it ("[]","{}"):
-Set: "{a,b,c}"
-List: "[a,b,c]"
-Dict: "{a:b,c:d}"
-Deque: "Deque(a,b,c)"
-LinkedList: "LinkedList(a,b,c)"
-Optional: "Optional(a)"

SIMD include it's DT:  
- repr -> "SIMD\[DType.float64, 1\]\(1.0\)"  

Optional, deque, LinkedList, List,.. don't include T: 
- repr -> "[0], {0}, {a:b}"

laszlokindrat reviewed (approved):
Thank you!
